drivers/sensor/tmag5273/tmag5273.c

```diff
@@ -0,0 +1,1230 @@
+			LOG_INF("VCC undervoltage detected");
+		}
+#if CONFIG_CRC
+		if (drv_cfg->crc_enabled &&
+		    (FIELD_GET(TMAG5273_OTP_CRC_ER_MSK, dev_stat) == TMAG5273_OTP_CRC_ERR)) {
+			LOG_INF("OTP CRC error detected");
+		}
+#endif
+		if (FIELD_GET(TMAG5273_INT_ER_MSK, dev_stat) == TMAG5273_INT_ERR) {
+			LOG_INF("INT pin error detected");
+		}
+
+		if (FIELD_GET(TMAG5273_OSC_ER_MSK, dev_stat) == TMAG5273_OSC_ERR) {
+			LOG_INF("Oscillator error detected");
+		}
+	}
+
+	return retval;
+}
+
+/**
+ * @brief sets the configuration of the interrupt gpio
+ *
+ * @param dev driver handle
+ * @param flags target configuration, or-ed to the flag from the dts file
+ *
+ * @return @see @ref gpio_pin_configure_dt
+ */
+static inline int tmag5273_int_gpio_config(const struct device *dev, const gpio_flags_t flags)
+{
+	const struct tmag5273_config *drv_cfg = dev->config;
+
+	int retval = gpio_pin_configure_dt(&drv_cfg->int_gpio, drv_cfg->int_gpio.dt_flags | flags);
+
+	if (retval < 0) {
+		LOG_ERR("cannot configure GPIO |%d|", retval);
+	}
+
+	return retval;
+}
+
+/**
+ * @brief performs a trigger through the INT-pin
+ *
+ * @param dev driver handle
+ *
+ * @retval 0 on success
+ * @retval see @ref gpio_pin_set_dt
+ */
+static inline int tmag5273_dev_int_trigger(const struct device *dev)
+{
+	int retval = tmag5273_int_gpio_config(dev, GPIO_OUTPUT);
+
+	if (retval < 0) {
+		return retval;
+	}
+
+	const struct tmag5273_config *drv_cfg = dev->config;
+
+	retval = gpio_pin_set_dt(&drv_cfg->int_gpio, 1);
+
+	if (retval >= 0) {
+		retval = gpio_pin_set_dt(&drv_cfg->int_gpio, 0);
+	}
+
+	if (retval < 0) {
+		LOG_ERR("error waking up through GPIO pin");
+	}
+
+	int retval_gpio = tmag5273_int_gpio_config(dev, GPIO_INPUT);
+
+	if (retval == 0) {
+		retval = retval_gpio;
+	}
+
+	return retval;
+}
+
+/**
+ * @brief update the measurement range of the X/Y/Z-axis
+ *
+ * @param dev handle to the sensor
+ * @param val value to be set
+ *
+ * @return see @ref i2c_reg_update_byte_dt
+ */
+static inline int tmag5273_attr_set_xyz_meas_range(const struct device *dev,
+						   const struct sensor_value *val)
+{
+	const struct tmag5273_config *drv_cfg = dev->config;
+	struct tmag5273_data *drv_data = dev->data;
+
+	const int16_t range_high = (drv_data->version == TMAG5273_VER_TMAG5273X1)
+					   ? TMAG5273_MEAS_RANGE_HIGH_MT_VER1
+					   : TMAG5273_MEAS_RANGE_HIGH_MT_VER2;
+	const int16_t range_low = (drv_data->version == TMAG5273_VER_TMAG5273X1)
+					  ? TMAG5273_MEAS_RANGE_LOW_MT_VER1
+					  : TMAG5273_MEAS_RANGE_LOW_MT_VER2;
+
+	const uint8_t regdata = (sensor_value_to_double(val) > range_low)
+					? TMAG5273_XYZ_MEAS_RANGE_HIGH
+					: TMAG5273_XYZ_MEAS_RANGE_LOW;
+
+	int retval = i2c_reg_update_byte_dt(&drv_cfg->i2c, TMAG5273_REG_SENSOR_CONFIG_2,
+					    TMAG5273_X_Y_RANGE_MSK | TMAG5273_Z_RANGE_MSK, regdata);
+
+	if (retval == 0) {
+		drv_data->xyz_range =
+			(regdata == TMAG5273_XYZ_MEAS_RANGE_LOW) ? range_low : range_high;
+	}
```

Comment:
changed this to only look for the integral part, since `val1` is sufficient to detect a "high" range. For the rest: I have problems here. The problem with this sensor is that it comes in different flavors, so I need (during runtime) figure out what the values which I need to use really are. I refactored it a little bit, put some stuff in functions to make it more readable, but more doesn't look possible from my perspective.